### PR TITLE
ansi-c: clang extensions

### PR DESCRIPTION
**Pull request by kroening**

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: aff0b394f5e08a6bcf9969751ae10694a5cbf683).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80288078
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: be6e3b6982f61c61c8408adf3f90550c272ede98).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80288229

**Comment by smowton** [comment]
Tests please

**Comment by kroening** [comment]
done

File: regression/ansi-c/gcc_float_types1/main.c [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+__float128 f128;
+#else
+__float80 f80;
+__float128 f128;
+#endif
+
+int main()
+{
+  #ifndef __clang__
+  // on gcc, they can be re-declared, as they are identifiers, not keywords
```
Comment (smowton) [inline regression/ansi-c/gcc_float_types1/main.c:15]: Perhaps run the test against Clang too, but checking for the expected syntax error in the .desc file?

File: src/util/config.cpp [hunk 2]
```diff
@@ -233,14 +234,15 @@ void configt::ansi_ct::set_arch_spec_power(const irep_idt &subarch)
...
     defines.push_back("__powerpc");
     defines.push_back("__powerpc__");
     defines.push_back("__POWERPC__");
     defines.push_back("__ppc__");
 
-    if(mode==flavourt::APPLE)
-      defines.push_back("__BIG_ENDIAN__");
+#if defined(__APPLE__)
+    defines.push_back("__BIG_ENDIAN__");
+#endif
```
Comment (tautschnig) [inline src/util/config.cpp]: This looks wrong to me: it introduces a compile-time configuration into what ought to be a runtime configuration choice. This needs to test for the (target) operating system (`os == "macos"`).

File: src/ansi-c/get-gcc-builtins.sh [hunk 3]
```diff
@@ -102,7 +102,7 @@ cat > builtins.h <<EOF
 // some newer versions of GCC apparently support __floatXYZ
 #define dfloat32_type_node __float32
 #define dfloat64_type_node __float64
-#define dfloat128_type_node __float128
+#define dfloat128_type_node __CPROVER_Float128
```
Comment (tautschnig) [inline src/ansi-c/get-gcc-builtins.sh:105]: Randomly placing this comment: the commit message says "bugfix" but it's not clear how this bug has been found/how a future regression is avoided?

File: src/ansi-c/scanner.l [hunk 4]
```diff
@@ -1389,6 +1389,14 @@ __decltype          { if(PARSER.cpp98 &&
                 return make_identifier();
             }
 
+  /* This is a clang extension */
```
Comment (tautschnig) [inline src/ansi-c/scanner.l]: Could a pointer to documentation be provided? Some day we may want to implement this.

File: src/ansi-c/ansi_c_internal_additions.cpp [hunk 5]
```diff
@@ -213,7 +213,17 @@ void ansi_c_internal_additions(std::string &code)
         code+="typedef long double __float128;\n";
     }
 
-    // On 64-bit systems, both gcc and clang have typedefs
+    if(
+      config.ansi_c.arch == "i386" || config.ansi_c.arch == "x86_64" ||
+      config.ansi_c.arch == "x32" || config.ansi_c.arch == "ia64")
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_internal_additions.cpp]: How has this subset of architectures been determined and/or why is it different from the `__float128` one?

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 3e1126936b2fc9ee6176408379028e6dcdefe53f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80377005

File: src/ansi-c/get-gcc-builtins.sh (hunk 3 above)
Comment (kroening) [reply src/ansi-c/get-gcc-builtins.sh:105]: This was found by running the existing regressions on FreeBSD, and repeating that should avoid it in the future as well.

File: src/util/config.cpp (hunk 2 above)
Comment (kroening) [reply src/util/config.cpp]: Ok, fixed.

File: regression/ansi-c/gcc_float_types1/main.c (hunk 1 above)
Comment (kroening) [reply regression/ansi-c/gcc_float_types1/main.c:15]: We don't have means to enable clang as preprocessor.

Comment (tautschnig) [reply regression/ansi-c/gcc_float_types1/main.c:15]: goto-cc --native-compiler=clang should do so, I believe?

**Review by tautschnig — APPROVED** [review]

File: regression/ansi-c/gcc_float_types1/main.c (hunk 1 above)
Comment (kroening) [reply regression/ansi-c/gcc_float_types1/main.c:15]: Yes; are we willing to add clang as a dependency for CI for this?

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: f74c161312ce8137c72c9f303c4a994d18f3a1b7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80402711

File: regression/ansi-c/gcc_float_types1/main.c (hunk 1 above)
Comment (tautschnig) [reply regression/ansi-c/gcc_float_types1/main.c:15]: I'd say it's more an "in principle we could" case. We'll probably do if/when we run into a regression on this :-)

**Pull request merged**